feat: thorchain lp ledger open app ack

gomesalexandre opened the pull request:
## Description

Does precisely what you'd expect it to do based on its box content, minus deposit/withdraw execution bits (implemented in `useSendThorTx` as part of https://github.com/shapeshift/web/pull/7628)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes https://github.com/shapeshift/web/issues/7580

## Risk

> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

- THORChain LP approval displays the Ledger Open App ack
- THORChain LP deposit displays the Ledger Open App ack
- THORChain LP withdraw displays the Ledger Open App ack

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

- ^

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

- ^

## Screenshots (if applicable)

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

- Approve

<img width="540" alt="Screenshot 2024-08-27 at 17 52 07" src="https://github.com/user-attachments/assets/d42df776-cb9c-4b9d-b593-34c05725e9f1">

- Add Liquidity

<img width="538" alt="Screenshot 2024-08-27 at 17 57 13" src="https://github.com/user-attachments/assets/f570dc8b-a07a-4c3c-92c9-76f06ccf6834">

- Remove Liquidity

<img width="633" alt="Screenshot 2024-08-27 at 18 06 02" src="https://github.com/user-attachments/assets/bd7d8449-9e5e-49e0-b078-f9927d6e9f07">


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on src/lib/utils/thorchain/hooks/useSendThorTx.tsx:
```diff
@@ -270,18 +270,18 @@ export const useSendThorTx = ({
     if (isToken(asset.assetId) && !inboundAddressData) return
 
+    await checkLedgerAppOpenIfLedgerConnected(asset.chainId)
+    if (accountNumber === undefined) return
```
`accountNumber` check needs to be last, we may not have an accountNumber (which is incorrect as we let users proceed, but another bug)

gomesalexandre replied:
https://github.com/shapeshift/web/issues/7634

kaladinlight replied:
I don't understand this change. Why would we care about checking if an app is open without the `accountNumber` to be able to continue anyways? Previous diff looks sane to me (only check if app is open just prior to build and broadcast)

gomesalexandre replied:
Just for the sake of completeness! Allows for it to be implemented everywhere for LP, even though the Tx won't go through withuot an accountNumber (as you've noticed, this is reverted in https://github.com/shapeshift/web/pull/7635)

kaladinlight commented on src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx:
```diff
@@ -873,7 +876,11 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
 
-  const handleApprove = useCallback(() => mutate(undefined), [mutate])
+  const handleApprove = useCallback(async () => {
+    if (!assetId) return
+    await checkLedgerAppOpenIfLedgerConnected(fromAssetId(assetId).chainId)
```
option to leverage `onMutate` to tie this logic directly to the mutation

gomesalexandre replied:
Prefer to keep it that way as that would make us destructure the `mutationFn` from `reactQueries.mutations.approve`  which looks a bit flaky IMO i.e:

```diff
diff --git a/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx b/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
index 9ceb939395..bcbe779ced 100644
--- a/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
+++ b/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
@@ -591,23 +591,30 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
     return serializeTxIndex(poolAssetAccountId, approvalTxId, poolAssetAccountAddress)
   }, [approvalTxId, poolAssetAccountAddress, poolAssetAccountId])
 
+  const { mutationKey, mutationFn } = reactQueries.mutations.approve({
+    assetId: poolAsset?.assetId,
+    spender: poolAssetInboundAddress,
+    amountCryptoBaseUnit: toBaseUnit(
+      actualAssetDepositAmountCryptoPrecision,
+      poolAsset?.precision ?? 0,
+      BigNumber.ROUND_UP,
+    ),
+    wallet: wallet ?? undefined,
+    from: poolAssetAccountId ? fromAccountId(poolAssetAccountId).account : undefined,
+    accountNumber: poolAssetAccountNumber,
+  })
+
   const {
     mutate,
     isPending: isApprovalMutationPending,
     isSuccess: isApprovalMutationSuccess,
   } = useMutation({
-    ...reactQueries.mutations.approve({
-      assetId: poolAsset?.assetId,
-      spender: poolAssetInboundAddress,
-      amountCryptoBaseUnit: toBaseUnit(
-        actualAssetDepositAmountCryptoPrecision,
-        poolAsset?.precision ?? 0,
-        BigNumber.ROUND_UP,
-      ),
-      wallet: wallet ?? undefined,
-      from: poolAssetAccountId ? fromAccountId(poolAssetAccountId).account : undefined,
-      accountNumber: poolAssetAccountNumber,
-    }),
+    mutationKey,
+    mutationFn: async (_: void) => {
+      if (!assetId) throw new Error('assetId is required')
+      await checkLedgerAppOpenIfLedgerConnected(fromAssetId(assetId).chainId)
+      return mutationFn(undefined)
+    },
     onSuccess: (txId: string) => {
       setApprovalTxId(txId)
     },
@@ -876,11 +883,7 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
     runeTxFeeCryptoBaseUnit,
   ])
 
-  const handleApprove = useCallback(async () => {
-    if (!assetId) return
-    await checkLedgerAppOpenIfLedgerConnected(fromAssetId(assetId).chainId)
-    mutate(undefined)
-  }, [assetId, checkLedgerAppOpenIfLedgerConnected, mutate])
+  const handleApprove = useCallback(() => mutate(undefined), [mutate])
 
   const handleSubmit = useCallback(() => {
     if (isApprovalRequired) return handleApprove()
```

kaladinlight replied:
Was thinking more along the lines of:

```
diff --git a/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx b/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
index 9ceb939395..81e9916893 100644
--- a/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
+++ b/src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx
@@ -608,6 +608,10 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
       from: poolAssetAccountId ? fromAccountId(poolAssetAccountId).account : undefined,
       accountNumber: poolAssetAccountNumber,
     }),
+    onMutate: async () => {
+      if (!poolAsset?.chainId) return
+      await checkLedgerAppOpenIfLedgerConnected(poolAsset.chainId)
+    },
     onSuccess: (txId: string) => {
       setApprovalTxId(txId)
     },
@@ -876,12 +880,6 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
     runeTxFeeCryptoBaseUnit,
   ])
 
-  const handleApprove = useCallback(async () => {
-    if (!assetId) return
-    await checkLedgerAppOpenIfLedgerConnected(fromAssetId(assetId).chainId)
-    mutate(undefined)
-  }, [assetId, checkLedgerAppOpenIfLedgerConnected, mutate])
-
   const handleSubmit = useCallback(() => {
     if (isApprovalRequired) return handleApprove()
     if (isSweepNeeded) return history.push(AddLiquidityRoutePaths.Sweep)
```

kaladinlight commented:
Looks like we are missing this check in `Sweep` which is a possible flow in Thorchain LP. Also noticing we simply haven't added this check around standard sends `useFormSend` as a side note

gomesalexandre commented:
> Looks like we are missing this check in `Sweep` which is a possible flow in Thorchain LP. Also noticing we simply haven't added this check around standard sends `useFormSend` as a side note

Actually very valid side note, added it in `handleSend` will both handle sweeps and regular sends - 39154e5e40b47fe39c34452d15b07b9c43282cc6:

- Add Liquidity Sweep

<img width="705" alt="image" src="https://github.com/user-attachments/assets/aaae53e2-6af8-4a06-bb08-2f9979f60220">

- THORChain Savers Deposit Sweep

<img width="637" alt="image" src="https://github.com/user-attachments/assets/3f0b8e0f-fbf4-42a6-9c50-c24b2fba7b36">


Note sends from the send modal are currently borked in this diff - to be tackled in a follow-up as well as we can't open a modal within a modal and we'll need to update the Modal provider to accomodate that.


Note the double checks in useSendThorTx's `executeTransaction`, which are expected for the time being (the first because of the current `accountNumber` shenanigans, fixed in https://github.com/shapeshift/web/pull/7635 above, the second being the *actual* one we want).
Going to follow-up with a holistic change at adapters-level so adapter methods that rely on on-device derivation take `checkLedgerAppOpenIfLedgerConnected` in instead, so this logic is better encapsulated and most importantly, so we are required to pass it.

kaladinlight commented on src/components/Modals/Send/utils.ts:
```diff
@@ -98,15 +98,21 @@ export const estimateFees = ({
   if (!asset) return ''
+
+  const chainId = asset.chainId
+  // The double check here and in broadcastTXID isn't a mistake - getAddress() does on-device address derivation
```
but... there is no double check there :P

gomesalexandre replied:
Nice catch haha, that was meant to be in `useSendThorTx`'s `executeTransaction`, however, removed to avoid merge conflicts in the next PR in the stack, and knowing we're going to cleanup and consolidate at implementation-level as a follow-up
93bf620d28

kaladinlight commented:
> Note the double checks in useSendThorTx's `executeTransaction`, which are expected for the time being (the first because of the current `accountNumber` shenanigans, fixed in #7635 above, the second being the _actual_ one we want). Going to follow-up with a holistic change at adapters-level so adapter methods that rely on on-device derivation take `checkLedgerAppOpenIfLedgerConnected` in instead, so this logic is better encapsulated and most importantly, so we are required to pass it.

On the same page with the holistic update. Current setup not super manageable and prone to regressions or forgetting on new implementations. Handling at adapter level makes a lot of sense. Wondering if we could leverage event emitters and have it completely encapsulated in chain adapters to reduce the need to propagate the function pass through everywhere. Then we can just listen for the ledger app open event and show the modal accordingly.
